File: Activities/ArticleImageActivity.cs
using System;
using System.IO;
using System.Threading;

using Android.App;
using Android.Content;
using Android.OS;
using Android.Widget;
using Android.Graphics;
using Android.Support.V4.Content;
using Android.Views;
using MatrixGuide;

using static VorratsUebersicht.Tools;

namespace VorratsUebersicht
{
    [Activity(Label = "ArticleImageActivity")]
    public class ArticleImageActivity : Activity
    {
        int articleId;
        bool editMode;
        ImageView imageView;
        TextView  imageInfo;
        Bitmap rotatedBitmap;
        bool isChanged = false;

        public static readonly int PickImageId = 1000;
        public static readonly int TakePhotoId = 1001;

        protected override void OnCreate(Bundle savedInstanceState)
        {
            base.OnCreate(savedInstanceState);

            SetContentView(Resource.Layout.ArticleImage);

            // ActionBar Hintergrund Farbe setzen
            var backgroundPaint = ContextCompat.GetDrawable(this, Resource.Color.Application_ActionBar_Background);
            backgroundPaint.SetBounds(0, 0, 10, 10);
            ActionBar.SetBackgroundDrawable(backgroundPaint);
            ActionBar.SetDisplayHomeAsUpEnabled(true);

            this.articleId = Intent.GetIntExtra    ("ArticleId", 0);
            this.editMode  = Intent.GetBooleanExtra("EditMode", false);
            string text    = Database.GetArticleName(this.articleId);

            this.imageView  = FindViewById<ImageView> (Resource.Id.ArticleImage_Image);
            this.imageView.Click += OnClickImage;

            this.imageInfo = FindViewById<TextView>(Resource.Id.ArticleImage_Info);

            ActionBar.Title = text;
            ActionBar.SetHomeButtonEnabled(true);
            ActionBar.SetIcon(Resource.Drawable.ic_photo_camera_white_24dp);

            if (ArticleDetailsActivity.imageLarge != null)
                this.ShowPictureFromBitmap();
            else
                this.ShowPictureFromDatabase();
        }

        private void OnClickImage(object sender, EventArgs e)
        {
            if (this.imageInfo.Visibility != ViewStates.Visible)
            {
                this.imageInfo.Visibility = ViewStates.Visible;
            }
            else
            {
                this.imageInfo.Visibility = ViewStates.Gone;
            }
        }

        public override bool OnCreateOptionsMenu(IMenu menu)
        {
            if (this.editMode)
            {
                MenuInflater.Inflate(Resource.Menu.ArticleImage_menu, menu);
            }
            return base.OnCreateOptionsMenu(menu);
        }

        public override bool OnOptionsItemSelected(IMenuItem item)
        {
            switch (item.ItemId)
            {
                case Resource.Id.ArticleImageMenu_RotateRight:
                    this.RotateImage();
                    return true;

                case Android.Resource.Id.Home:
                    this.OnBackPressed();
                    return true;
            }

            return base.OnOptionsItemSelected(item);
        }

        public override void OnBackPressed()
        {
            if (!this.isChanged)
            {
                base.OnBackPressed();
                return;
            }

            var progressDialog = this.CreateProgressBar();
            new Thread(new ThreadStart(delegate             
            {
                this.SaveBitmap();
                GC.Collect();
                
                RunOnUiThread(() => base.OnBackPressed());

                this.HideProgressBar(progressDialog);
            })).Start();
        }

        private void SaveBitmap()
        {
            if (this.rotatedBitmap == null)
                return;

            if (!this.isChanged)
                return;

            MemoryStream stream = new MemoryStream();
            this.rotatedBitmap.Compress(Bitmap.CompressFormat.Png, 100, stream);
            ArticleDetailsActivity.imageLarge = stream.ToArray();


            byte[] resizedImage = ImageResizer.ResizeImageAndroid(this.rotatedBitmap, 48*2, 85*2);

            Bitmap smallBitmap = BitmapFactory.DecodeByteArray(resizedImage, 0, resizedImage.Length);

            stream = new MemoryStream();
            smallBitmap.Compress(Bitmap.CompressFormat.Png, 100, stream);
            ArticleDetailsActivity.imageSmall = stream.ToArray();

            Intent intent = new Intent();
            this.SetResult(Result.Ok, intent);

            this.rotatedBitmap = null;
            this.isChanged = false;
        }

        private void ShowPictureFromDatabase()
        {
            if (this.articleId == 0)
            {
                this.imageView.SetImageResource(Resource.Drawable.ic_photo_camera_white_24dp);
                this.imageView.Alpha = 1f;
                return;
            }

            ArticleImage article = Database.GetArticleImage(this.articleId);
            if (article == null)
            {
                this.imageView.SetImageResource(Resource.Drawable.ic_photo_camera_black_24dp);
                this.imageView.Alpha = 0.5f;

                return;
            }

            if (article.ImageLarge == null)
            {
                this.imageView.SetImageResource(Resource.Drawable.ic_photo_camera_black_24dp);
                this.imageView.Alpha = 0.5f;

                return;
            }

            string message;
            try
            {
                Bitmap largeBitmap = BitmapFactory.DecodeByteArray (article.ImageLarge, 0, article.ImageLarge.Length);

                this.imageView.SetImageBitmap(largeBitmap);

                message = string.Format("Bild (BxH): {0:n0} x {1:n0} (Größe: {2:n0}, Komprimiert: {3:n0})\r\n", 
                    largeBitmap.Width,
                    largeBitmap.Height,
                    Tools.ToFuzzyByteString(largeBitmap.ByteCount),
                    Tools.ToFuzzyByteString(article.ImageLarge.Length));


                Bitmap smallBitmap = BitmapFactory.DecodeByteArray (article.ImageSmall, 0, article.ImageSmall.Length);

                message  += string.Format("Thn. (BxH): {0:n0} X {1:n0} (Größe: {2:n0}, Komprimiert: {3:n0})",
                    smallBitmap.Width,
                    smallBitmap.Height,
                    Tools.ToFuzzyByteString(smallBitmap.ByteCount),
                    Tools.ToFuzzyByteString(article.ImageSmall.Length));

                largeBitmap = null;
                smallBitmap = null;
            }
            catch (Exception e)
            {
                TRACE(e);

                message = e.Message;
                this.imageView.SetImageResource(Resource.Drawable.baseline_error_outline_black_24);
                this.imageView.Alpha = 1f;
            }               

            this.imageInfo.Text = message;
            article = null;
        }

        private void ShowPictureFromBitmap()
        {
            Bitmap largeBitmap = BitmapFactory.DecodeByteArray(ArticleDetailsActivity.imageLarge, 0, ArticleDetailsActivity.imageLarge.Length);
            this.rotatedBitmap = largeBitmap;
            this.imageView.SetImageBitmap(largeBitmap);

            string message;

            message = string.Format("Bild (BxH): {0:n0} x {1:n0} (Größe: {2:n0}, Komprimiert: {3:n0})\r\n", 
                largeBitmap.Width,
                largeBitmap.Height,
                Tools.ToFuzzyByteString(largeBitmap.ByteCount),
                Tools.ToFuzzyByteString(ArticleDetailsActivity.imageLarge.Length));

            if (ArticleDetailsActivity.imageSmall != null)
            {
                Bitmap smallBitmap = BitmapFactory.DecodeByteArray(ArticleDetailsActivity.imageSmall, 0, ArticleDetailsActivity.imageSmall.Length);

                message  += string.Format("Thn. (BxH): {0:n0} X {1:n0} (Größe: {2:n0}, Komprimiert: {3:n0})",
                    smallBitmap.Width,
                    smallBitmap.Height,
                    Tools.ToFuzzyByteString(smallBitmap.ByteCount),
                    Tools.ToFuzzyByteString(ArticleDetailsActivity.imageSmall.Length));
            }

            this.imageInfo.Text = message;
        }

        private void RotateImage()
        {
            if (this.rotatedBitmap == null)
            {
                ArticleImage article = Database.GetArticleImage(this.articleId, true);
                if (article == null)
                    return;

                if (article.ImageLarge == null)
                    return;
                try
                {
                    this.rotatedBitmap = BitmapFactory.DecodeByteArray (article.ImageLarge, 0, article.ImageLarge.Length);
                    GC.Collect();
                }
                catch(Exception ex)
                {
                    TRACE(ex);

                    this.rotatedBitmap = null;
                    this.imageView.SetImageResource(Resource.Drawable.baseline_error_outline_black_24);
                    this.imageView.Alpha = 1f;
                    this.imageInfo.Text = ex.Message;
                    return;
                }
            }

            try
            {
                // Bild um 90 Grad drehen
                Matrix mat = new Matrix();
                mat.PostRotate(90);
                Bitmap bMapRotate = Bitmap.CreateBitmap(this.rotatedBitmap, 0, 0, this.rotatedBitmap.Width, this.rotatedBitmap.Height, mat, true);
                this.imageView.SetImageBitmap(bMapRotate);
                this.rotatedBitmap = bMapRotate;
                this.isChanged = true;

                string message = string.Format("Bild (BxH): {0:n0} X {1:n0} (Größe: {2:n0})", 
                    this.rotatedBitmap.Width, 
                    this.rotatedBitmap.Height, 
                    Tools.ToFuzzyByteString(this.rotatedBitmap.ByteCount));

                this.imageInfo.Text = message;

            }
            catch(Exception ex)
            {
                TRACE(ex);

                this.rotatedBitmap = null;
                this.imageView.SetImageResource(Resource.Drawable.baseline_error_outline_black_24);
                this.imageView.Alpha = 1f;
                this.imageInfo.Text = ex.Message;
            }
        }

        private ProgressBar CreateProgressBar()
        {
            var progressBar = FindViewById<ProgressBar>(Resource.Id.ArticleImage_ProgressBar);
            progressBar.Visibility = ViewStates.Visible;
            this.Window.SetFlags(WindowManagerFlags.NotTouchable, WindowManagerFlags.NotTouchable);
            return progressBar;
        }

        private void HideProgressBar(ProgressBar progressBar)
        {
            RunOnUiThread(() =>
            {
                progressBar.Visibility = ViewStates.Invisible;
                this.Window.ClearFlags(WindowManagerFlags.NotTouchable);
            });
        }
   }
}

File: app/src/main/java/de/stryi/vorratsuebersicht/ArticleImageActivity.kt
package de.stryi.vorratsuebersicht

import android.annotation.SuppressLint
import android.content.Intent
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.Matrix
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import android.view.View
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.view.menu.MenuBuilder
import androidx.lifecycle.lifecycleScope
import de.stryi.vorratsuebersicht.databinding.ArticleImageBinding
import de.stryi.vorratsuebersicht.database.Database
import de.stryi.vorratsuebersicht.tools.Tools
import de.stryi.vorratsuebersicht.tools.toPngByteArray
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.util.Locale


class ArticleImageActivity : AppCompatActivity() {

    private lateinit var binding: ArticleImageBinding

    var articleId: Int = 0
    var editMode: Boolean = false
    var isChanged: Boolean = false

    var rotatedBitmap: Bitmap? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        this.binding = ArticleImageBinding.inflate(layoutInflater)

        setContentView(binding.root)

        this.setSupportActionBar(binding.ArticleImageAppBar)

        this.articleId = intent.getIntExtra("ArticleId", 0)
        this.editMode  = intent.getBooleanExtra("EditMode", false)

        if (intent.hasExtra("Title"))
        {
            this.title = intent.getStringExtra("Title")
        }
        else
        {
            this.title = Database.getArticleName(this.articleId)
        }

        binding.ArticleImageAppBar.overflowIcon?.setTint(getColor(R.color.Application_ActionBar_TextColor))
        binding.ArticleImageAppBar.setNavigationOnClickListener { finish() }
        binding.ArticleImageAppBar.setOnMenuItemClickListener { this.onOptionsItemSelected(it) }
        binding.ArticleImageImage.setOnClickListener    { this.showImageInformation() }
        binding.ArticleImageImageThn.setOnClickListener { this.showImageInformation() }

        if (ArticleDetailsActivity.imageLarge != null)
        {
            this.showPictureFromBitmap()
        }
        else
        {
            this.showPictureFromDatabase()
        }
    }

    @SuppressLint("RestrictedApi")
    override fun onCreateOptionsMenu(menu: Menu): Boolean
    {
        if (menu is MenuBuilder) {
            menu.setOptionalIconsVisible(true)
        }
        menuInflater.inflate(R.menu.article_image_menu, menu)
        return true
    }

    override fun onPrepareOptionsMenu(menu: Menu): Boolean
    {
        if (!this.editMode)
        {
            val itemRotate = menu.findItem(R.id.ArticleImage_Menu_RotateRight)
            itemRotate.isVisible = false
        }

        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.ArticleImage_Menu_RotateRight -> {
                this.rotateImage()
                return true
            }
        }

        return false
    }

    override fun finish()
    {
        if (!this.isChanged)
        {
            super.finish()
            return
        }

        this.showProgressBar()
        lifecycleScope.launch {
            withContext(Dispatchers.IO) {
                saveBitmap()
                System.gc()
            }

            finish()
            hideProgressBar()
        }
    }

    private fun saveBitmap()
    {
        if (rotatedBitmap == null)
            return

        // Großes Bild als PNG-Bytearray speichern
        ArticleDetailsActivity.imageLarge = rotatedBitmap!!.toPngByteArray()

        // Verkleinertes Bild (Thumbnail) erstellen
        val smallBitmap = ImageResizer.resizeImageAndroid(
            rotatedBitmap!!,
            (48 * 2).toFloat(),
            (85 * 2).toFloat())

        // Thumbnail als PNG-Bytearray speichern
        ArticleDetailsActivity.imageSmall = smallBitmap

        val intent = Intent()
        this.setResult(RESULT_OK, intent)

        this.rotatedBitmap = null
        this.isChanged = false
    }

    fun showPictureFromDatabase()
    {
        val article= Database.getArticleImage(this.articleId)

        this.showPictureFromByteArray(article?.imageLarge, article?.imageSmall)
    }

    fun showPictureFromBitmap()
    {
        this.showPictureFromByteArray(ArticleDetailsActivity.imageLarge, ArticleDetailsActivity.imageSmall)
    }

    fun showPictureFromByteArray(imageLarge: ByteArray?, imageSmall: ByteArray?)
    {
        if (imageLarge == null)
        {
            binding.ArticleImageImage.setImageResource(R.drawable.hide_image_24px)
            binding.ArticleImageImage.visibility = View.VISIBLE
            binding.ArticleImageInfo.text = resources.getString(R.string.ArticleImage_SizeLarge)
        }

        if (imageSmall == null)
        {
            binding.ArticleImageImageThn.setImageResource(R.drawable.hide_image_24px)
            binding.ArticleImageImageThn.visibility = View.VISIBLE
            binding.ArticleImageInfoThn.text = resources.getString(R.string.ArticleImage_SizeSmall)
        }

        try
        {
            if (imageLarge != null)
            {
                val largeBitmap = BitmapFactory.decodeByteArray(imageLarge, 0, imageLarge.size)
                binding.ArticleImageImage.setImageBitmap(largeBitmap)

                this.rotatedBitmap = largeBitmap

                val message = String.format(
                    Locale.getDefault(),
                    "Bild (BxH): %,d x %,d\nGröße: %s, Komprimiert: %s",
                    largeBitmap.width,
                    largeBitmap.height,
                    Tools.toFuzzyByteString(largeBitmap.byteCount.toLong()),
                    Tools.toFuzzyByteString(imageLarge.size.toLong()))

                binding.ArticleImageInfo.text = message
            }

            if (imageSmall != null)
            {
                val smallBitmap = BitmapFactory.decodeByteArray(imageSmall, 0, imageSmall.size)
                binding.ArticleImageImageThn.setImageBitmap(smallBitmap)

                val message = String.format(
                    Locale.getDefault(),
                    "Vorschaubild (BxH): %,d x %,d\nGröße: %s, Komprimiert: %s",
                    smallBitmap.width,
                    smallBitmap.height,
                    Tools.toFuzzyByteString(smallBitmap.byteCount.toLong()),
                    Tools.toFuzzyByteString(imageSmall.size.toLong()))

                binding.ArticleImageInfoThn.text = message
            }
        }
        catch (e: Exception)
        {
            binding.ArticleImageInfoThn.text = e.message
            binding.ArticleImageInfoThn.visibility = View.VISIBLE
        }
    }

    private fun rotateImage()
    {
        if (rotatedBitmap == null)
        {
            return
        }

        val matrix = Matrix()
        matrix.postRotate(90f)

        val rotated = Bitmap.createBitmap(
            rotatedBitmap!!,
            0, 0,
            rotatedBitmap!!.width,
            rotatedBitmap!!.height,
            matrix,
            true
        )

        binding.ArticleImageImage.setImageBitmap(rotated)
        rotatedBitmap = rotated
        this.isChanged = true
    }

    private fun showImageInformation()
    {
        if (binding.ArticleImageInfo.visibility != View.VISIBLE)
        {
            binding.ArticleImageInfo.visibility     = View.VISIBLE
            binding.ArticleImageInfoThn.visibility  = View.VISIBLE
            binding.ArticleImageImageThn.visibility = View.VISIBLE
        }
        else
        {
            binding.ArticleImageInfo.visibility     = View.GONE
            binding.ArticleImageInfoThn.visibility  = View.GONE
            binding.ArticleImageImageThn.visibility = View.GONE
        }
    }

    private fun showProgressBar()
    {
        binding.ArticleImageProgressBar.visibility = View.VISIBLE
    }

    private fun hideProgressBar()
    {
        binding.ArticleImageProgressBar.visibility = View.GONE
    }
}